pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/ingress_reconciler.go

```diff
@@ -311,6 +320,88 @@ func createIngress(isvc *v1beta1.InferenceService, config *v1beta1.IngressConfig
 		hosts = append(hosts, serviceHost)
 		gateways = append(gateways, config.IngressGateway)
 	}
+	if config.UrlTemplate != "" {
+		url, err := GenerateUrl(isvc.Name, isvc.Namespace, config)
+		if err != nil {
+			log.Error(err, "Failed to generate URL from urlTemplate")
+			return nil
+		}
+		url.Path = strings.TrimSuffix(url.Path, "/") // remove trailing "/" if present
+		if url.Path != "" {
+			// In this case, we have a path-based URL so we add a path-based rule
+			httpRoutes = append(httpRoutes, &istiov1alpha3.HTTPRoute{
+				Match: []*istiov1alpha3.HTTPMatchRequest{
+					{
+						Uri: &istiov1alpha3.StringMatch{
+							MatchType: &istiov1alpha3.StringMatch_Prefix{
+								Prefix: url.Path + "/",
+							},
+						},
+						Authority: &istiov1alpha3.StringMatch{
+							MatchType: &istiov1alpha3.StringMatch_Regex{
+								Regex: constants.HostRegExp(url.Host),
+							},
+						},
+						Gateways: []string{config.IngressGateway},
+					},
+					{
+						Uri: &istiov1alpha3.StringMatch{
+							MatchType: &istiov1alpha3.StringMatch_Exact{
+								Exact: url.Path,
+							},
+						},
+						Authority: &istiov1alpha3.StringMatch{
+							MatchType: &istiov1alpha3.StringMatch_Regex{
+								Regex: constants.HostRegExp(url.Host),
+							},
+						},
+						Gateways: []string{config.IngressGateway},
+					},
+				},
+				Rewrite: &istiov1alpha3.HTTPRewrite{
+					Uri: "/",
+				},
+				Route: []*istiov1alpha3.HTTPRouteDestination{
+					createHTTPRouteDestination(backend, isvc.Namespace, config.LocalGatewayServiceName),
+				},
+				Headers: &istiov1alpha3.Headers{
+					Request: &istiov1alpha3.Headers_HeaderOperations{
+						Set: map[string]string{
+							"Host": network.GetServiceHostname(backend, isvc.Namespace),
+						},
+					},
+				},
+			})
+		} else if !contains(hosts, url.Host) {
+			// In this case, we have a host-based URL, so we add a host-based rule
```

Comment:
I am confused about this case, in which scenario the `url.Host` is not in the `hosts` list?

Reply:
Well this is a corner case where we have set a host-based url template:
```
urlTemplate: https://{{ .Name }}-{{ .Namespace}}.serving.mydomain.com/
```
This is more or less the case where we configure KServe to expose the ISVC and let KNative with default values (example.com domain).